pinot-server/src/main/java/com/linkedin/pinot/server/starter/helix/HelixServerStarter.java

```diff
@@ -164,9 +167,32 @@ public HelixServerStarter(String helixClusterName, String zkServer, Configuratio
     serverMetrics.addCallbackGauge("memory.allocationFailureCount", PinotDataBuffer::getAllocationFailureCount);
   }
 
-  private void updateInstanceConfigInHelix(int adminApiPort, boolean shuttingDown) {
+  private void waitForAllSegmentsLoaded() {
+    if (_helixServerConfig.getBoolean(CommonConstants.Server.CONFIG_OF_STARTER_ENABLE_SEGMENTS_LOADING_CHECK, CommonConstants.Server.DEFAULT_STARTER_ENABLE_SEGMENTS_LOADING_CHECK)) {
+      long startTime = System.currentTimeMillis();
+      int serverStarterTimeout = _helixServerConfig.getInt(CommonConstants.Server.CONFIG_OF_STARTER_TIMEOUT_IN_SECONDS, CommonConstants.Server.DEFAULT_STARTER_TIMEOUT_IN_SECONDS);
+      boolean allSegmentsLoaded = false;
+      while (System.currentTimeMillis() - startTime < 1000 * serverStarterTimeout) {
+        if (ServiceStatus.getServiceStatus() == Status.GOOD) {
+          LOGGER.info("All the segments are fully loaded into Pinot server.");
+          allSegmentsLoaded = true;
+          break;
+        }
+        try {
+          LOGGER.info("Waiting for all segments loaded, sleep 10 seconds...");
+          Thread.sleep(10000);
+        } catch (Exception e) {
+          // Do nothing
+        }
+      }
+      if (!allSegmentsLoaded) {
+        LOGGER.info("Segments are not fully loaded within {} seconds...", serverStarterTimeout);
+      }
+    }
+  }
+
+  private void setAdminApiPort(int adminApiPort) {
     Map<String, String> propToUpdate = new HashMap<>();
-    propToUpdate.put(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS, String.valueOf(shuttingDown));
```

Comment:
Shouldn't remove this flag. Otherwise Pinot broker won't be notified that Server has already been back up.